### PR TITLE
Bugfix #8103: scipy.stats.boxcox - np.nan produces warnings

### LINKED ISSUE #8103: scipy.stats.boxcox - np.nan produces warnings, but it shouldn't
I would suggest practical enhancement to the **scipy.stats.boxcox** method.

Currently, this method is not able to handle np.nan values nicely - and produces full console of warnings. 
Having np.nan in data is common thing and it is logically and gracefully handled by many similar methods in scientific computing - for example:

- np.log(..) - when input is np.nan, then output is also np.nan. Clear, simple and makes sense.

This is logical, consistent a one would expect the same behavior also in the _boxcox_ method.

### Reproducing code example:
```python
import numpy as np
from scipy import stats

# Learn lambda (for example on train-set, to apply it later on test-set)
_, lmbda = stats.boxcox(np.array([1, 2, 2, 2, 3, 3, 3, 4, 5, 6]))  

# Apply learnt lambda - here the PROBLEM occurs
transformed = stats.boxcox(np.array([1, 2, np.nan, 2, 3, 3, 3, 4, 5, 6]), lmbda=lmbda)
```

### Error message:
Problem is, that code above produces strange warning:
```
_Anaconda3\lib\site-packages\scipy\stats\morestats.py:1030: RuntimeWarning: invalid value encountered in less_equal
  if any(x <= 0):_
```
This bomethod is callled many times in data-science projects, this warning fills up whole console with hundreds warning. 
Putting warning here for np.nan does not bring much value, and gracefully making it np.nan would be logical (the same way, as it works for np.log() function)

### Scipy/Numpy/Python version information:

<<Output from 'import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)'>>
```
0.19.1 1.13.1 sys.version_info(major=3, minor=6, micro=2, releaselevel='final', serial=0)
```

### Expectated correct behavior

- np.nan values should be consistently accepted the same way, as it is in np.log(..) function
- i.e. no warning should be thrown, in case np.nan is present

### PR DESCRIPTION
**Pull request by stefansimik**
Closes #8103

### CODE REVIEW

File: scipy/stats/morestats.py [hunk 1]
```diff
@@ -1027,14 +1031,22 @@ def boxcox(x, lmbda=None, alpha=None):
...
-    if any(x <= 0):
+    contains_nan, nan_policy = _contains_nan(x, nan_policy)
+    not_nan_mask = ~np.isnan(x)
+
+    if any(x[not_nan_mask] <= 0):
         raise ValueError("Data must be positive.")
 
     if lmbda is not None:  # single transformation
-        return special.boxcox(x, lmbda)
+        return special.boxcox(x, lmbda)  # propagates nan's
```
Comment (person142) [inline scipy/stats/morestats.py]: I think it should be

```python
if nan_policy == 'propagate':
    return special.boxcox(x, lmbda)
else:
    return special.boxcox(x[not_nan_mask], lmbda)
```

It won't broadcast `lmbda` correctly, but that doesn't seem like an intended use case. @josef-pkt or @ev-br would know better than me though, so we should see what they say.

Comment (stefansimik) [reply scipy/stats/morestats.py]: Great finding, thank you.
To achieve the same, I just moved code for omitting nan values.
Check please, if it meets your expectations.

**Comment by person142** [comment]
Looks good to me, but I forgot to ask you to add a regression test (my bad). Would probably be fine to take your original example from https://github.com/scipy/scipy/issues/8103 and check that you get the right results with the various possible nan policies.

**Comment by josef-pkt** [comment]
about broadcasting: given that the docstring says x is 1-dim, it won't be available.
e.g. taking the mask indexed array would ravel.

I think an extra "omit-fit" nan-policy would be useful in this case, which is now not available, AFAICS.
That is, omit the nans in the estimation of lmbda, but propagate the nans to the transformed array.

(E.g. that would make it easier to combine the transformed array with some original data like a pandas DataFrame).

**Comment by josef-pkt** [comment]
The docstring is missing a See Also to the scipy.special function, which users can use if they don't need to estimate lambda and have broadcasting.

**Comment by stefansimik** [comment]
@person142 Tests were added, but I don't know how to run them locally on my Windows machine.
I expect, it requires some special setup to compile whole scipy first, install it, and so on...
If you know some tutorial how to do it locally, let me know.

Otherwise, I will wait until CI runs and checks all tests.

**Comment by stefansimik** [comment]
@josef-pkt 
My understanding is, that 

- `lmbda `be calculated only using real values i.e. `nan `values are not usable in calculating lmbda itself.
- in this context - `nan_policy `parameter logically applies only to the transformation of the values - and not calculation of `lmbda `-- but here, I agree, this should be more clearly stated in the docs and I would be happy to do it, if you agree

**Comment by josef-pkt** [comment]
@stefansimik but AFAICS, you don't have an option for computing lmbda without nans, but returning the array with nans. Or do I misread the changes?

**Comment by stefansimik** [comment]
@josef-pkt

Yes, that's true - you are right. That can happen = computing lmbda without nan's, but returning transformed values with nan's.

In my understanding, the calculation of lambda itself is not supposed to be parametrized by nan_policy. But maybe it should be, I am open, if you can see some practical use-case.

I can see some value in consistency, if we added additional param lmbda_nan_policy, that would have default value 'omit', and other possible value 'raise'. Option 'propagate' is useless here.

But I have other suggestion to fix and answer all these questions. I will write it here in new comment from my PC...

**Comment by stefansimik** [comment]
Guys, I would like to suggest different direction - let's try to imagine, that boxcox method will no new parameter(s) like nan_policy, but propagates nan's always.

How it will work and handle all possible scenarios:

- **nan propagation scenario:** this is default and most common behavior.
- **nan omit scenario:**  --> one can remove nan's if this is requirement from application point of view
- **nan raise scenario:**  --> one can check nan's and raise exception is requirement from application point of view

What are main benefits:

- simple, straightforward and mainly intuitive approach
  - the same way, like we can log-transform, we can boxcox-transform
- less parameters, less complexity, less thinking
- less maintenance

On top of that, I believe, that function like np.log(), or boxcox():

- should do its core logic and return results. Nothing more, nothing less.
  - that means, such methods should not add too much on top of basic functionality = it should not do complex checking of inputs, nor support multiple input check scenarios to meet various. expectations of different domains. This is responsibility of applications, which uses boxcox() method to make sure it prepares proper inputs - and this is based on application domain knowledge.

Do you know anybody, who would require to put some nan_policy into np.log(...) method ?

So let's get rid of the nan_policy. As we dont need nan_policy in np.log(...), the same way we don't need nan_policy in boxcox().

If one sends array with nan's into boxcox:

- then lmbda will be calculated from values other than nan's
- and transformation will be done on values also other than nan's
- where nan's is, there it remains.

Let's make it work the same way as np.log(...)
Let's make it practically working, simple and intuitive - as Python is expected to be.

I like simplicity. I like to have things intuitive.
So that's my suggestion - but you guys have the last word of course ;)

**Comment by josef-pkt** [comment]
@stefansimik There is a difference between scipy.stats and scipy.special.
The scipy special functions are elementwise and behave like you describe,
The scipy.stats functions need to do some further processing over the entire array in almost all cases. So the analogy to np.log is not appropriate and e.g. np.mean versus np.nanmean would be a better analogy.

In this case, stats.boxcox as a simple transformation is now essentially obsolete given the scipy.special function. The advantage is that we estimate and transform at the same time, as a convenience function that combines two separate steps. 
Similar in hypothesis tests or most other stats functions we have to let the user decide how to handle nan values, because simply propagating them throughout is not possible.

(Note: my initial suggestion was to use 'ignore' instead of 'propagate' as an option to not even check for nans if checking for nans is relatively expensive for a function that itself does just some simple computation. I'm slowly starting to change my mind about whether that's a useful default if the function has to do more than some very simple computation.)

**Comment by stefansimik** [comment]
@josef-pkt
Thank you for detailed explanation.

> Similar in hypothesis tests or most other stats functions we have to let the user decide how to handle nan values, because simply propagating them throughout is not possible.

I can understand, that in scipy.stats, there  hypothesis tests / other stats functions, where nan_policy is  relevant and nature of the problem makes it useful.

But are you sure, this is case also for scipy.stats.boxcox(..) ?
I mean, boxcox(..) without nan_policy looks as better and transparent choice.

As stats.boxcox() does 2 thing at once:

 - **calculation of lambda** has to be done by omitting nan's - always. 
  - There is no other option = nothing to configure. It just need to be documented in 1-2 sentences.
- **transformation** can be done with nan's - by propagating nan's. 
  - Why to add options for throwing exceptions here, or omit nan's, when everyone can do it itself - in special case, they really want. They can just remove nan's manually before calling transformation.

So I can understand, that there are cases in scipy.stats, where nan_police is good,
I am not sure, this is case for boxcox(...)

- Reason is, that there are no known scenarios to me, where nan_policy would be useful in the stats.boxcox(). 
- But I see strong negative against the nan_policy:
  - Added complexity does not bring anything significantly useful, that couldn't be done otherwise, nor anything that would compensate for more complex behavior, documentation, testing and maintenance.

Most usages of stats.boxcox() will be straightforward = people want get results for boxcox() and continue with their workflow, instead of raising / catching exceptions. Anyway - in rare cases, someone wants to raise, he is open to do it.  So why to incorporate such logic into the core of the function?

I see it similarly for 'omit' option, because one can write one line -  `x[~is.nan(x)]` - and remove nan's before calling boxcox() easily - in special case, it is required by application logic. It is trivial to do so I don't see the practical value of incorporating such functionality into the core of the function.

I am open-minded but it looks to me like, there is no significant added value in nan_policy parameter for the this case of stats.boxcox() method, that would compensate for all the added complexity.

**Comment by stefansimik** [comment]
@josef-pkt @person142 
Hi guys, any comments or suggestions? What do you prefer?
Your opinion is important now, so we can finalize this issue and make scipy even better ;)

**Comment by person142** [comment]
I will defer to whatever @josef-pkt thinks, since he is much wiser in the ways of `scipy.stats` than I am.

**Comment by stefansimik** [comment]
@person142 @josef-pkt 
Thank you. I agree and will do the same.
Despite any opinions, I am still learning scipy so I will be happy to update 
code based on @josef-pkt decision.
Seriously, no more comments, just implementation of final solution. :)

@josef-pkt could you help us to resolve this issue, based on your opinion, please?

**Comment by ilayn** [comment]
@josef-pkt when you have time, could you please have a look at this?

**Comment by josef-pkt** [comment]
I still have problems with this, and I'm still not sure what the appropriate solution should be.

The underlying problem also affects other functions, e.g.
https://github.com/scipy/scipy/blob/master/scipy/stats/stats.py#L2246
(zscore, zmap, obrientransform and possibly the trimming functions are the current elementwise functions, besides box-cox if lambda is estimated.)

Those function are elementwise like ufuncs where nan propagation is a useful and the usual default. However, they need intermediate results where a single nan destroys the results.

The "obvious" (most common usage) solution would be to behave like ufuncs and always propagate nan in the elementwise part, but use nan robust computation in the reduce statistics.
But I'm not sure what the options for this should be, and whether decisions should be hard coded instead of optional. 
The most useful option IMO would be `raise` and `fit-omit`. AFAIK, all current usage for nan-policy are for reduce functions. So, there is no established pattern for ufunc_like functions that also need a reduce.

The trimming functions are a bit in between, because they explicitly exclude some elements by definition.

For box-cox a similar issue as nan handling would be what to do with non-positive values, options would be to drop or add a constant to make them positive. That is also left to the user, and the function currently hardcodes "raise".


aside: I just saw that sigmaclip is missing in the function list at
https://github.com/scipy/scipy/blob/master/scipy/stats/stats.py#L87

**Pull request closed without merging**